승자 확인하기

TTOzzi opened the pull request:
1. 클래스를 분리하고 속성들에 private 을 설정하다보니 게임 외부에서 플레이어의 카드에 접근하기위해 클로저를 PokerGame 에서 Players 로 (PokerGame.forEachPlayers), Players 에서 각 플레이어로(Players.forEach), 또 플레이어에서 각각의 카드로(Player.forEachCard, Hand.forEach) 전달해주는 형태가 되었습니다. 좋은 방법이 아닌것같은데 다른 방법이 없을까요?

2. 점수를 저장할 위치를 많이 고민했는데 Player 에 두려니 Card 를 탐색하러 들어가야하고 Hand 에 두려니 Player 에서 꺼내와야했습니다. 결국 Hand 에 두고 꺼내오는 방식으로 구현했는데 이것을 클래스로 따로 분리하는게 더 좋을까요?

TTOzzi committed:
PokerGame 클래스에서 Players 클래스 분리

TTOzzi committed:
Model 관련 파일들 Model 폴더로 이동

TTOzzi committed:
전체 파일 접근제어 설정

TTOzzi committed:
makeGame 에서 playerStack 을 만드는부분 분리

TTOzzi committed:
Hand 클래스 생성

TTOzzi committed:
함수 이름 의미에 맞게 변경

TTOzzi committed:
스트레이트 확인을 위한 함수, 같은 카드 갯수 확인을 위한 함수 추가

스트레이트는 중복 카드를 없애고 연속된 카드의 갯수를 확인
같은 카드 확인은 카드배열을 dictionary 로 맵핑해서 생성

TTOzzi committed:
점수 계산을 위한 함수 구현

set 을 활용해 스트레이트를 먼저 확인하고
dictionary 로 같은 카드의 갯수 확인

TTOzzi committed:
승자 표시를 위한 속성 추가

TTOzzi committed:
게임이 끝난 후 승자 계산 구현

TTOzzi committed:
승자 표시를 위한 이미지 추가

TTOzzi committed:
Score 의 rawValue 에 직접 접근하지 않고 대소비교를 하도록 수정

TTOzzi committed:
winnerIndex 를 모델에서 계산해서 내보내도록 수정

TTOzzi committed:
승자 표시 화면출력 구현

게임을 뷰에 추가 한 뒤 승자의 출력위치 옆에 승자표시가 출력되도록 구현

TTOzzi committed:
점수 확인 함수 분리, 스코어와 숫자들을 따로 반환하도록 수정

스코어가 같을때, 높은 숫자를 비교하기위해 분리
만약 스코어가 1페어로 같고 1페어에 해당하는 숫자도 같을 때 다음숫자를 비교하기위해 배열로 숫자들을 반환해줌
다음숫자도 같다면 그 다음숫자를 비교하도록 해줌

TTOzzi committed:
플레이어 점수 비교를 위한 Equatable 프로토콜 채택

TTOzzi committed:
기존 승자표시 삭제 로직 개선

godrm commented on CardGameApp/CardGameApp/Model/Card.swift:
```diff
@@ -17,6 +17,9 @@ class Card {
     private var rank: Rank
+    var number: Int {
+        rank.rawValue
+    }
```
값을 가져가서 비교하려고 하지말고, Card 자체를 비교하는 `<`, `==` 연산자를 구현하세요.
값을 가져가는 로직을 보면 늘 제 말을 기억하세요.

godrm commented on CardGameApp/CardGameApp/Model/Player.swift:
```diff
@@ -0,0 +1,51 @@
+
+class Player {
+    private var hand = Hand()
+    var isWinner = false
```
여러 인스턴스가 만들어지는 경우에 그 중에 하나를 선택하는 방식을 이렇게 isWinner 속성을 활용하는 것은
오동작할 경우가 많아집니다. 여러 인스턴가 모두 isWinner = true 상태일 수 있으니까요
심지어 public 으로 하면 Player 속성을 외부에서 바꿀 수 있는거니까 좋지 않습니다.

godrm commented on CardGameApp/CardGameApp/Model/Player.swift:
```diff
@@ -0,0 +1,51 @@
+                    continue
+                } else {
+                    return lhs.numbers[index] > rhs.numbers[index]
+                }
```
hand를 점수 계산하는 역할과 책임을 하도록 만들었으면 hand에서 number를 가져와서 여기서 비교하지마세요.
대신 Hand 타입이 그 역할을 다 하고, Hand 자체를 비교하면 되지 않을까요? 
그게 객체 지향 프로그래밍에서 객체를 협력하도록 만드는 방식입니다. 여기서는 lhs.hand 와 rhs.hand를 비교하면 됩니다.

godrm commented on CardGameApp/CardGameApp/Model/Players.swift:
```diff
@@ -0,0 +1,35 @@
+    
+    func whoIsWinner() {
+        let result = list.sorted { $0 > $1 }
+        result[0].isWinner = true
```
winner를 player.isWinner를 넣어놓을 게 아니라 winner를 return 해주는 방식으로 구현하는 게 어떨까요?

godrm commented on CardGameApp/CardGameApp/Model/PokerGame.swift:
```diff
@@ -39,43 +39,51 @@ class PokerGame {
+            }
+            index += 1
+        }
+        return winnerIndex
```
위에 winner 처리하는 방식이 바뀌면 이 부분도 불필요할 겁니다.

TTOzzi committed:
winner 처리 방식 변경

TTOzzi committed:
Hand 클래스 Equatable 프로토콜 채택

플레이어를 비교할때 Hand 내부 속성으로 비교하는것이 아닌 Hand 자체를 비교하도록 수정

TTOzzi committed:
Card 에서 rank 의 rawValue 를 직접 가져오는게 아닌 Card 를 직접 비교하도록 수정

다음 Rank 를 알기위한 nextRank
다음 카드를 반환해주는 nextCard()
Card 를 Dictionary 에 담아주기위한 Hashable 프로토콜

TTOzzi committed:
checkStraight 로직 수정

정렬을 내림차순으로 해서 nextCard 를 index 에 써줘야하는데 index - 1 에 쓰고있었음 -> 수정
guard 로 옵셔널을 벗겨냄

TTOzzi committed:
checkStraight 로직 수정

continue 시 index 를 안더해줘서 무한루프를 돔 -> 수정

TTOzzi committed:
checkStraight 로직 수정

이전 커밋으로 했어야했는데 잘못보냈습니다.

TTOzzi commented:
피드백 수정 완료했습니다!
최대한 로직을 유지하면서 값을 직접 가져오지않도록 수정하기위해 잘 쓰지않던 프로토콜과 lazy var 등을 사용하였는데 좀 더 프로토콜에 대해 잘 알게된것같습니다.

godrm commented on CardGameApp/CardGameApp/Model/Card.swift:
```diff
@@ -14,22 +14,33 @@ class Card {
+        var nextRank: Rank? {
+            guard let nextRank = Rank.init(rawValue: self.rawValue + 1) else { return nil }
+            return nextRank
+        }
```
이거 보다는 isNextRank(with: Rank) -> Bool 이런 메소드가 있으면 어떨까요? 
값을 가져가는 게 아니라 가져와서 비교하는 방식을 고려하세요.

godrm commented on CardGameApp/CardGameApp/Model/Card.swift:
```diff
@@ -14,22 +14,33 @@ class Card {
+    func nextCard() -> Card? {
+        guard let nextRank = Rank.init(rawValue: self.rank.rawValue + 1) else { return nil }
+        return Card(shape: self.shape, rank: nextRank)
+    }
```
Card도 값을 가져가는 게 아니라 가져와서 비교하는 방식을 고려하는 게 좋습니다.

godrm commented on CardGameApp/CardGameApp/Model/Hand.swift:
```diff
@@ -33,21 +34,28 @@ class Hand {
             index += 1
         }
-        return straightCount == 5
+        return false
```
위에 피드백 한것처럼 index가 아니라 객체 자체에 접근해서 비교하는 방법을 고려해보세요.

godrm commented on CardGameApp/CardGameApp/Model/Hand.swift:
```diff
@@ -91,8 +98,27 @@ class Hand {
         let info = getInfo()
-        return info.map { (value) -> Int in value.key }
+        return info.map { (value) -> Card in value.key }
+    }
```
이걸 가져가서 비교하는 곳이 있는데 
탐색해서 결과를 넘겨주는 메소드가 있고, 그 값을 비교하는 형태는 어떨까요?

godrm commented:
비교하는 부분들이 많이 추상화되어 간단해졌는데, 
여전히 배열을 가져가서 비교하는 방식이 남아있습니다. 
머지할테니 여유가 되면 개선해보는 것도 좋겠습니다. 
수고하셨습니다.

TTOzzi commented:
말씀해주신 내용들 반영해서 수정해보겠습니다.
감사합니다!!
